KomaMRIFiles/src/Sequence/WritePulseq.jl

```diff
@@ -0,0 +1,463 @@
+    for (obj, _) in gradunique_obj_id
+        shape_amp = obj.A / magsign(obj.A)
+        if all([
+            !(
+                length(shape_amp) == length(shape_amp_unique) &&
+                shape_amp ≈ shape_amp_unique
+            ) for (shape_amp_unique, _) in gradunique_amp_id
+        ])
+            push!(gradunique_amp_id, [shape_amp, id_shape_cnt])
+            id_shape_cnt = id_shape_cnt + 1
+        end
+        if isa(obj.T, Vector{<:Number})
+            shape_tim = cumsum([0; obj.T]) / seq.DEF["GradientRasterTime"]
+            if all([
+                !(
+                    length(shape_tim) == length(shape_tim_unique) &&
+                    shape_tim ≈ shape_tim_unique
+                ) for (shape_tim_unique, _) in gradunique_tim_id
+            ])
+                push!(gradunique_tim_id, [shape_tim, id_shape_cnt])
+                id_shape_cnt = id_shape_cnt + 1
+            end
+        end
+    end
+    return gradunique_amp_id, gradunique_tim_id, id_shape_cnt
+end
+
+"""
+    compressed_data = compress(data::Vector)
+
+Returns the compressed data vector `compressed_data` for the intupt vector `data` using the
+algorithm for encoding the derivative in a run-length compressed format.
+"""
+function compress(data)
+    return compress_shape(data)[2]
+end
+
+"""
+    write_seq(seq::Sequence, filename::String)
+
+Writes a .seq file for a given sequence `seq` y the location `filename`
+"""
+function write_seq(seq::Sequence, filename)
+    # Get the unique objects (RF, Grad y ADC) and its IDs
+    rfunique_obj_id = get_typeunique_obj_id(get_typeon_obj(seq, "rf"))
+    grunique_obj_id = get_typeunique_obj_id(get_typeon_obj(seq, "gr"))
+    adcunique_obj_id = get_typeunique_obj_id(get_typeon_obj(seq, "adc"))
+    gradunique_obj_id = [[obj, id] for (obj, id) in grunique_obj_id if length(obj.A) != 1]
+    trapunique_obj_id = [[obj, id] for (obj, id) in grunique_obj_id if length(obj.A) == 1]
+    rfunique_abs_id, rfunique_ang_id, rfunique_tim_id, id_shape_cnt = get_rfunique(
+        rfunique_obj_id, 1, seq
+    )
+    gradunique_amp_id, gradunique_tim_id, _ = get_gradunique(
+        gradunique_obj_id, id_shape_cnt, seq
+    )
+    # Define the table to be written for the [BLOCKS] section
+    @warn "EXTENSIONS will not be handled"
+    rf_id = [id for (_, _, id) in get_type_blk_obj_id(seq, "rf", rfunique_obj_id)]
+    gx_id = [id for (_, _, id) in get_type_blk_obj_id(seq, "gx", grunique_obj_id)]
+    gy_id = [id for (_, _, id) in get_type_blk_obj_id(seq, "gy", grunique_obj_id)]
+    gz_id = [id for (_, _, id) in get_type_blk_obj_id(seq, "gz", grunique_obj_id)]
+    adc_id = [id for (_, _, id) in get_type_blk_obj_id(seq, "adc", adcunique_obj_id)]
+    blk_obj_dur_rf_gx_gy_gz_adc_ext = [
+        [b, s, 0, rf_id[b], gx_id[b], gy_id[b], gz_id[b], adc_id[b], 0] for
+        (b, s) in enumerate(seq)
+    ]
+    for bodrxyzae in blk_obj_dur_rf_gx_gy_gz_adc_ext
+        blk = bodrxyzae[1]
```

Comment:
Variable names.

Reply:
Names improved in  2dd30b0 bcf5515 511dc91 72718e6